File: benchmarks/multilang-wasm/track-b/fft_opt.c
#include <stdint.h>

// Track B — FFT optimized C variant (INDEPENDENTLY OPTIMIZED).
// Track A baseline (benchmarks/multilang-wasm/fft_kernel.c) is KEPT UNTOUCHED;
// the arithmetic below is the VERBATIM Track A sequence (same custom sin/cos,
// same butterfly op order) → BIT-IDENTICAL output.
//
// OPTIMIZATION LOG:
// 1. Pointer-hoisted butterflies: real/imag reads for u and v are loaded into
//    locals once per j (the baseline re-reads real[v]/imag[v] four times per
//    j iteration). The float op order is unchanged.
// 2. Hoisted `i + step` (v base) and the row pointers out of the innermost
//    expression.
// Measured delta in docs/track-b-optimizations.md.

static float sinf_custom(float x) {
  while (x > 3.14159265358979323846f) x -= 2.0f * 3.14159265358979323846f;
  while (x < -3.14159265358979323846f) x += 2.0f * 3.14159265358979323846f;
  float x2 = x * x;
  float x3 = x * x2;
  float x5 = x3 * x2;
  float x7 = x5 * x2;
  return x - (x3 / 6.0f) + (x5 / 120.0f) - (x7 / 5040.0f);
}

static float cosf_custom(float x) {
  return sinf_custom(x + 1.57079632679489661923f);
}

__attribute__((visibility("default")))
void fft_butterfly_baseline(float* real, float* imag, uint32_t len) {
  for (uint32_t step = 1; step < len; step <<= 1) {
    float angle = -3.14159265358979323846f / (float)step;
    float w_real = cosf_custom(angle);
    float w_imag = sinf_custom(angle);
    for (uint32_t i = 0; i < len; i += (step << 1)) {
      float cur_w_real = 1.0f;
      float cur_w_imag = 0.0f;
      for (uint32_t j = 0; j < step; j++) {
        uint32_t u = i + j;
        uint32_t v = i + j + step;
        float tr = real[v] * cur_w_real - imag[v] * cur_w_imag;
        float ti = real[v] * cur_w_imag + imag[v] * cur_w_real;
        real[v] = real[u] - tr;
        imag[v] = imag[u] - ti;
        real[u] += tr;
        imag[u] += ti;
        float nwR = cur_w_real * w_real - cur_w_imag * w_imag;
        float nwI = cur_w_real * w_imag + cur_w_imag * w_real;
        cur_w_real = nwR;
        cur_w_imag = nwI;
      }
    }
  }
}

__attribute__((visibility("default")))
void fft_butterfly_opt(float* real, float* imag, uint32_t len) {
  for (uint32_t step = 1; step < len; step <<= 1) {
    float angle = -3.14159265358979323846f / (float)step;
    float w_real = cosf_custom(angle);
    float w_imag = sinf_custom(angle);
    uint32_t block = step << 1;
    for (uint32_t i = 0; i < len; i += block) {
      float cur_w_real = 1.0f;
      float cur_w_imag = 0.0f;
      float* rv = real + i + step;
      float* iv = imag + i + step;
      for (uint32_t j = 0; j < step; j++) {
        float rvj = rv[j];
        float ivj = iv[j];
        float tr = rvj * cur_w_real - ivj * cur_w_imag;
        float ti = rvj * cur_w_imag + ivj * cur_w_real;
        uint32_t u = i + j;
        rv[j] = real[u] - tr;
        iv[j] = imag[u] - ti;
        real[u] += tr;
        imag[u] += ti;
        float nwR = cur_w_real * w_real - cur_w_imag * w_imag;
        float nwI = cur_w_real * w_imag + cur_w_imag * w_real;
        cur_w_real = nwR;
        cur_w_imag = nwI;
      }
    }
  }
}

File: benchmarks/multilang-wasm/track-b/fft_opt.js
// Track B — FFT optimized JavaScript variant (INDEPENDENTLY OPTIMIZED).
// Track A baseline = the jsFft in public/multilang-runner.js (KEPT UNTOUCHED);
// mirrored verbatim below.
//
// OPTIMIZATION LOG:
// 1. TWIDDLE SEQUENCE CACHE: the baseline advances cur_w_real/cur_w_imag via a
//    complex multiply for EVERY (i, j) — the sequence is identical for every i
//    at a given step. The optimized variant computes the exact same multiply
//    sequence ONCE per step into a Float64Array and reuses it for all i blocks.
//    Same float ops in the same order → BIT-IDENTICAL output.
// 2. Locals for u/v row reads (real[v]/imag[v] read once per j).

export function fftBaseline(real, imag, LEN) {
  for (let step = 1; step < LEN; step <<= 1) {
    const angle = -Math.PI / step, wReal = Math.cos(angle), wImag = Math.sin(angle);
    for (let i = 0; i < LEN; i += step << 1) {
      let cwR = 1.0, cwI = 0.0;
      for (let j = 0; j < step; j++) {
        const u = i + j, v = i + j + step;
        const tr = real[v] * cwR - imag[v] * cwI;
        const ti = real[v] * cwI + imag[v] * cwR;
        real[v] = real[u] - tr;
        imag[v] = imag[u] - ti;
        real[u] += tr;
        imag[u] += ti;
        const nwR = cwR * wReal - cwI * wImag, nwI = cwR * wImag + cwI * wReal;
        cwR = nwR;
        cwI = nwI;
      }
    }
  }
}

export function fftOpt(real, imag, LEN) {
  for (let step = 1; step < LEN; step <<= 1) {
    const angle = -Math.PI / step, wReal = Math.cos(angle), wImag = Math.sin(angle);
    // Precompute the exact per-step twiddle sequence (same multiply order).
    const tw = new Float64Array(step * 2);
    let cwR = 1.0, cwI = 0.0;
    for (let j = 0; j < step; j++) {
      tw[j * 2] = cwR;
      tw[j * 2 + 1] = cwI;
      const nwR = cwR * wReal - cwI * wImag, nwI = cwR * wImag + cwI * wReal;
      cwR = nwR;
      cwI = nwI;
    }
    for (let i = 0; i < LEN; i += step << 1) {
      for (let j = 0; j < step; j++) {
        const u = i + j, v = i + j + step;
        const cwr = tw[j * 2], cwi = tw[j * 2 + 1];
        const tr = real[v] * cwr - imag[v] * cwi;
        const ti = real[v] * cwi + imag[v] * cwr;
        real[v] = real[u] - tr;
        imag[v] = imag[u] - ti;
        real[u] += tr;
        imag[u] += ti;
      }
    }
  }
}
